src/binary_annotator.cpp

```diff
@@ -0,0 +1,768 @@
+
+    fields_processed++;
+  });
+
+  // Check if we covered all the expectant fields. If not, we need to add them
+  // as unknown fields.
+  const uint16_t expectant_vtable_fields =
+      (vtable_size - sizeof(uint16_t) - sizeof(uint16_t)) / sizeof(uint16_t);
+
+  for (uint16_t id = fields_processed; id < expectant_vtable_fields; ++id) {
+    const uint64_t field_offset = offset_start + id * sizeof(uint16_t);
+    const uint16_t offset_from_table = GetScalar<uint16_t>(field_offset);
+
+    VTable::Entry entry;
+    entry.field = nullptr;  // No field to reference.
+    entry.offset_from_table = offset_from_table;
+    fields.insert(std::make_pair(id, entry));
+
+    regions.push_back(MakeBinaryRegion(
+        field_offset, sizeof(uint16_t), BinaryRegionType::VOffset, 0, 0,
+        std::string("offset to unknown field (id: " + std::to_string(id) +
+                    ")")));
+  }
+
+  sections_[vtable_offset] = MakeBinarySection(
+      table->name()->str(), BinarySectionType::VTable, std::move(regions));
+
+  VTable vtable;
+  vtable.fields = std::move(fields);
+  vtable.table_size = table_size;
+  vtable.vtable_size = vtable_size;
+
+  vtables_[vtable_offset] = vtable;
+}
+
+void BinaryAnnotator::BuildTable(uint64_t offset, const BinarySectionType type,
+                                 const reflection::Object *table) {
+  std::vector<BinaryRegion> regions;
+  const uint64_t table_offset = offset;
+
+  // Tables start with the vtable
+  const uint64_t vtable_offset = table_offset - GetScalar<int32_t>(offset);
+  regions.emplace_back(
+      MakeBinaryRegion(offset, sizeof(int32_t), BinaryRegionType::SOffset, 0,
+                       vtable_offset, std::string("offset to vtable")));
+  offset += sizeof(int32_t);
+
+  // Parse the vtable first so we know what the rest of the fields in the table
+  // are.
+  BuildVTable(vtable_offset, table);
+
+  const VTable &vtable = vtables_.at(vtable_offset);
+
+  // This is the size and length of this table.
+  const uint16_t table_size = vtable.table_size;
+  const uint64_t table_end_offset = table_offset + table_size;
+
+  const uint64_t field_offset_start = offset;
+
+  // We need to iterate over the vtable fields by their offset in the binary,
+  // not by their IDs. So copy them over to another vector that we can sort on
+  // the offset_from_table property.
+  std::vector<VTable::Entry> fields;
+  for (const auto &vtable_field : vtable.fields) {
+    fields.push_back(vtable_field.second);
+  }
+
+  std::stable_sort(fields.begin(), fields.end(),
+                   [](const VTable::Entry &a, const VTable::Entry &b) {
+                     return a.offset_from_table < b.offset_from_table;
+                   });
+
+  // Iterate over all the fields by order of their offset.
+  for (size_t i = 0; i < fields.size(); ++i) {
+    const reflection::Field *field = fields[i].field;
+    const uint16_t offset_from_table = fields[i].offset_from_table;
+
+    if (!offset_from_table) {
```

Comment:
optional nit: might want to avoid implicit conversions `offset_from_table == 0`

actually, tbh we're pretty inconsistent about this already 🤷

Reply:
Done. No, i think its good to be consistent, and that starts at each PR :)